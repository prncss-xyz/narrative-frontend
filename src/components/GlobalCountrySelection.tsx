import { createContext, useContext, useState } from "react";
import { Country } from "../hooks/countries";
import { setInArray } from "../utils/arrays";
import { Box, Flex, H3 } from "./basics";
import { ToggleButton } from "./ToggleButton";
import { TogglingSelector } from "./TogglingSelector";

type CountryCodes = string[];

const Context = createContext<
  null | [null | CountryCodes, (state: CountryCodes) => void]
>(null);

export function countryString(
  countries: Country[],
  activeCountryCodes: string[]
) {
  const activeCountryNames = countries
    .filter(({ countryCode }) => activeCountryCodes.includes(countryCode))
    .map(({ name }) => name)
    .sort();
  if (activeCountryNames.length === 0) return "none";
  if (activeCountryNames.length === 1) return activeCountryNames[0];
  const last = activeCountryNames.at(-1);
  const rest = activeCountryNames.slice(0, -1);
  return [rest.join(", "), last].join(" & ");
}

export function CountrySelectorContext({
  children,
}: {
  children: React.ReactNode;
}) {
  const [activeCountryCodes, setActiveCountryCodes] = useState<null | string[]>(
    null
  );
  return (
    <Context.Provider value={[activeCountryCodes, setActiveCountryCodes]}>
      {children}
    </Context.Provider>
  );
}

export function useGlobalCountyList(
  countries: Country[]
): [CountryCodes, (state: CountryCodes) => void] {
  const context = useContext(Context);
  if (!context)
    throw new Error(
      "CountrySelector should always be used inside CountrySelectorContext"
    );
  // eslint-disable-next-line prefer-const
  let [activeCountryCodes, setActiveCountryCodes] = context;
  activeCountryCodes ??= countries.map((country) => country.countryCode);
  return [activeCountryCodes, setActiveCountryCodes];
}

export function sortCountriesByName(countries: Country[]) {
  return [...countries].sort((a, b) =>
    a.name < b.name ? -1 : a.name > b.name ? 1 : 0
  );
}

export function GlobalCountrySelector({
  countries,
  ...props
}: {
  countries: Country[];
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  countries = sortCountriesByName(countries);
  const [activeCountryCodes, setActiveCountryCodes] =
    useGlobalCountyList(countries);
  return (
    <Box>
      <H3>Included countries</H3>
      <Flex gap={3} {...props}>
        {countries.map((country) => (
          <div key={country.countryCode}>
            <ToggleButton
              active={activeCountryCodes.includes(country.countryCode)}
              setActive={(status) =>
                // status
                setActiveCountryCodes(
                  // activeCountryCodes
                  setInArray(activeCountryCodes, country.countryCode, status)
                )
              }
            >
              {country.name}
            </ToggleButton>
          </div>
        ))}
      </Flex>
    </Box>
  );
}

export function GlobalCountrySelectionSummary({
  count,
  countries,
  ...props
}: {
  count: number;
  countries: Country[];
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  countries = sortCountriesByName(countries);
  const [activeCountryCodes, setActiveCountryCodes] =
    useGlobalCountyList(countries);
  return (
    <Flex
      justifyItems="start"
      flexWrap="wrap"
      color="muted"
      alignItems="baseline"
      pb={2}
      textAlign="left"
      columnGap="0.4em"
      {...props}
    >
      <Box display="inline">Showing</Box>
      <Box display="inline" color="text" fontWeight="bold">
        {count}
      </Box>
      <Box display="inline">{count === 1 ? "result" : "results"} from</Box>
      <Box display="inline" color="text" fontWeight="bold">
        {countryString(countries, activeCountryCodes)}
      </Box>
    </Flex>
  );
}
