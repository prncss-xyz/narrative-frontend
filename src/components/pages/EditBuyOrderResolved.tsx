import { useState } from "react";
import { Navigate } from "react-router-dom";
import {
  BuyOrder,
  BuyOrderSchema,
  useUpdateBuyOrder,
} from "../../hooks/buyOrders";
import { Country } from "../../hooks/countries";
import { Dataset } from "../../hooks/datasets";
import { ActionBox, Flex } from "../basics";
import { BuyOrderForm, FormBuyOrder } from "../BuyOrderForm";
import { Clickable } from "../Clickable";

function Actions({ buyOrder }: { buyOrder: BuyOrder }) {
  const mutation = useUpdateBuyOrder();
  return (
    <>
      {mutation.data && <Navigate to="/buy-order-list" />}
      <Flex justifyContent="center">
        <Clickable onClick={() => mutation.mutate(buyOrder)}>
          <ActionBox>Save</ActionBox>
        </Clickable>
      </Flex>
    </>
  );
}

export function EditBuyOrderResolved({
  buyOrder,
  datasets,
  countries,
}: {
  buyOrder: BuyOrder;
  datasets: Dataset[];
  countries: Country[];
}) {
  const [formBuyOrder, setFormBuyOrder]: [
    FormBuyOrder,
    (s: FormBuyOrder) => void
  ] = useState<FormBuyOrder>(buyOrder);
  return (
    <BuyOrderForm
      buyOrder={formBuyOrder}
      setBuyOrder={setFormBuyOrder}
      datasets={datasets}
      countries={countries}
    >
      <Actions buyOrder={BuyOrderSchema.parse(formBuyOrder)} />
    </BuyOrderForm>
  );
}
