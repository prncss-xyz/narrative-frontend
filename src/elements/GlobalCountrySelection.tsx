import { Flex, Box, H3 } from "./basics";
import { Overlay } from "./Overlay";
import { Clickable } from "./Clickable";
import { createContext, useContext, useState } from "react";
import { Country } from "../hooks/countries";
import { RoundedButton } from "./RoundedButton";
import { TogglingSelector } from "./TogglingSelector";

type CountryCodes = string[];

const Context = createContext<
  null | [null | CountryCodes, (state: CountryCodes) => void]
>(null);

export function countryString(countries: Country[], activeCountryCodes: string[]) {
  const activeCountryNames = countries
    .filter(({ countryCode }) => activeCountryCodes.includes(countryCode))
    .map(({ name }) => name)
    .sort();
  if (activeCountryNames.length === 0) return "none";
  if (activeCountryNames.length === 1) return activeCountryNames[0];
  const last = activeCountryNames.at(-1);
  const rest = activeCountryNames.slice(0, -1);
  return [rest.join(", "), last].join(" & ");
}

export function CountrySelectorContext({
  children,
}: {
  children: React.ReactNode;
}) {
  const [activeCountryCodes, setActiveCountryCodes] = useState<null | string[]>(null);
  return (
    <Context.Provider value={[activeCountryCodes, setActiveCountryCodes]}>
      {children}
    </Context.Provider>
  );
}

export function useGlobalCountyList(
  countries: Country[]
): [CountryCodes, (state: CountryCodes) => void] {
  const context = useContext(Context);
  if (!context)
    throw new Error(
      "CountrySelector should always be used inside CountrySelectorContext"
    );
  let [activeCountryCodes, setActiveCountryCodes] = context;
  activeCountryCodes ??= countries.map((country) => country.countryCode);
  return [activeCountryCodes, setActiveCountryCodes];
}

export function sortCountriesByName(countries: Country[]) {
  return [...countries].sort((a, b) =>
    a.name < b.name ? -1 : a.name > b.name ? 1 : 0
  );
}

export function GlobalCountryList({
  count,
  countries,
}: {
  count: number;
  countries: Country[];
}) {
  const [overlayVisible, setOverlayVisible] = useState(false);
  countries = sortCountriesByName(countries);
  const [activeCountryCodes, setActiveCountryCodes] = useGlobalCountyList(countries);
  return (
    <>
      <Overlay
        visible={overlayVisible}
        onClickOutside={() => setOverlayVisible(false)}
      >
        <Box>
          <H3>Included countries</H3>
          <Flex gap={3}>
            <TogglingSelector
              state={activeCountryCodes}
              setState={setActiveCountryCodes}
              items={countries.map((country) => ({
                key: country.countryCode,
                toElem: (props) => (
                  <RoundedButton {...props}>{country.name}</RoundedButton>
                ),
              }))}
            />
          </Flex>
        </Box>
      </Overlay>
      <Clickable onClick={() => setOverlayVisible(true)}>
        <Flex color="gray1">
          <Box>Showing</Box>
          <Box color="black" fontWeight="bold">
            &nbsp;{count}&nbsp;
          </Box>
          <Box>results from</Box>
          <Box color="black" fontWeight="bold">
            &nbsp;{countryString(countries, activeCountryCodes)}&nbsp;
          </Box>
        </Flex>
      </Clickable>
    </>
  );
}
