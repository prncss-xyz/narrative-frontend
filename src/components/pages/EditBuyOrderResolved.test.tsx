import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { CountrySelectorContext } from "../GlobalCountrySelection";
import { EditBuyOrderResolved } from "./EditBuyOrderResolved";

export const countries = [
  {
    countryCode: "US",
    name: "United States",
    storedData: [
      { datasetId: 1, recordCount: 500 },
      { datasetId: 2, recordCount: 0 },
      { datasetId: 3, recordCount: 2500 },
      { datasetId: 4, recordCount: 750 },
      { datasetId: 5, recordCount: 50 },
      { datasetId: 6, recordCount: 1000 },
    ],
  },
  {
    countryCode: "GB",
    name: "United Kingdom",
    storedData: [
      { datasetId: 1, recordCount: 0 },
      { datasetId: 2, recordCount: 1250 },
      { datasetId: 3, recordCount: 2000 },
      { datasetId: 4, recordCount: 1500 },
      { datasetId: 5, recordCount: 500 },
      { datasetId: 6, recordCount: 500 },
    ],
  },
];

export const datasets = [
  {
    id: 1,
    name: "fishing",
    label: "Fishing Zones",
    description:
      "Data about global fishing zones, including yields, historical trends and ocean life surveys.",
    thumbnailUrl: "https://picsum.photos/id/1038/92/92",
    costPerRecord: 0.03,
  },
  {
    id: 2,
    name: "shipping",
    label: "Shipping Routes",
    description:
      "Metadata regarding global shipping routes. Contains data such as traffic volumne, operating restrictiona and mileage of routes.",
    thumbnailUrl: "https://picsum.photos/id/1002/92/92",
    costPerRecord: 0.05,
  },
  {
    id: 3,
    name: "ports",
    label: "Ports of Entry",
    description: "Detailed information about ports related to cargo shipping.",
    thumbnailUrl: "https://picsum.photos/id/1026/92/92",
    costPerRecord: 0.02,
  },
  {
    id: 4,
    name: "ocean",
    label: "Ocean Measurements",
    description:
      "Detailed measurements of many ocean locations, including LAT/LON, temperature, humidity and precipitation.",
    thumbnailUrl: "https://picsum.photos/id/1024/92/92",
    costPerRecord: 0.01,
  },
  {
    id: 5,
    name: "weather",
    label: "Weather and Climate",
    description: "Historical averages and trends of oceanographic climate.",
    thumbnailUrl: "https://picsum.photos/id/135/92/92",
    costPerRecord: 0.04,
  },
];

const buyOrderData = {
  id: "1",
  createdAt: new Date("2022-10-25T21:09:34+0000"),
  name: "My first buy order",
  datasetIds: [1, 3, 4],
  countries: ["US", "GB"],
  budget: 2500,
};

describe("EditBuyOrderResolved", () => {
  it("should render correctly", () => {
    const queryClient = new QueryClient();
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <CountrySelectorContext>
          <Router>
            <Routes>
              <Route
                path="/"
                element={
                  <EditBuyOrderResolved
                    countries={countries}
                    datasets={datasets}
                    buyOrder={buyOrderData}
                  />
                }
              ></Route>
            </Routes>
          </Router>
        </CountrySelectorContext>
      </QueryClientProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
