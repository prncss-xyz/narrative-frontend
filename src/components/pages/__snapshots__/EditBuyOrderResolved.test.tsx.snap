// Vitest Snapshot v1

exports[`EditBuyOrderResolved > should render correctly 1`] = `
<div>
  <div
    class="css-9wihcd-Box-Flex e2i8k6n2"
  >
    <div
      class="css-1dv1990-Box-Flex e2i8k6n2"
      width="6"
    >
      <div
        class="css-uoxz8q-Box-Grid e2i8k6n1"
      >
        <div
          class="css-1x3sxzl-Box e2i8k6n3"
        >
          <h3
            class="css-5jl3ge-Box e2i8k6n3"
            color="muted"
            font-size="1"
            font-weight="normal"
            text-decoration="underline"
          >
            Order name
          </h3>
          <input
            class="css-1ld7d19-InputRaw e2i8k6n0"
            placeholder="name"
            type="text"
            value="My first buy order"
            width="100%"
          />
        </div>
        <div
          class="css-1x3sxzl-Box e2i8k6n3"
        >
          <h3
            class="css-5jl3ge-Box e2i8k6n3"
            color="muted"
            font-size="1"
            font-weight="normal"
            text-decoration="underline"
          >
            Date Created
          </h3>
          <div
            class="css-fqccl6-Box e2i8k6n3"
            height="1"
          >
            10/25/2022
          </div>
        </div>
        <div
          class="css-1x3sxzl-Box e2i8k6n3"
        >
          <h3
            class="css-5jl3ge-Box e2i8k6n3"
            color="muted"
            font-size="1"
            font-weight="normal"
            text-decoration="underline"
          >
            Order budget
          </h3>
          <div
            class="css-1pkfq1d-Box-Flex e2i8k6n2"
          >
            <div
              class="css-1x3sxzl-Box e2i8k6n3"
            >
              $
            </div>
            <input
              class="css-wt16d4-InputRaw e2i8k6n0"
              placeholder="budget"
              type="text"
              value="2500"
              width="100%"
            />
          </div>
        </div>
        <div
          class="css-1x3sxzl-Box e2i8k6n3"
        >
          <h3
            class="css-5jl3ge-Box e2i8k6n3"
            color="muted"
            font-size="1"
            font-weight="normal"
            text-decoration="underline"
          >
            Forecasted Records
          </h3>
          <div
            class="css-fqccl6-Box e2i8k6n3"
            height="1"
          >
            7250 of 9050 available records
          </div>
        </div>
      </div>
      <div
        class="css-1x3sxzl-Box e2i8k6n3"
      >
        <h3
          class="css-5jl3ge-Box e2i8k6n3"
          color="muted"
          font-size="1"
          font-weight="normal"
          text-decoration="underline"
        >
          Included datasets
        </h3>
        <div
          class="css-7v5rcp-Box-Grid e2i8k6n1"
        >
          <div>
            <div
              class="css-p155cz-Box-Flex e2i8k6n2"
            >
              <img
                alt="dataset thumbnail"
                class="css-p7u6kb-Img e2i8k6n4"
                height="1"
                src="https://picsum.photos/id/1038/92/92"
              />
              <div
                class="css-d9aoz6-Box-Flex e2i8k6n2"
              >
                <div
                  class="css-9g59n2-Box e2i8k6n3"
                  color="text"
                  font-size="2"
                >
                  Fishing Zones
                </div>
                <div
                  class="css-11jaf6e-Box e2i8k6n3"
                  color="muted"
                  font-size="0"
                >
                  $
                  0.03
                   per record
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="css-svkabz-Box-Flex e2i8k6n2"
            >
              <img
                alt="dataset thumbnail"
                class="css-p7u6kb-Img e2i8k6n4"
                height="1"
                src="https://picsum.photos/id/1002/92/92"
              />
              <div
                class="css-d9aoz6-Box-Flex e2i8k6n2"
              >
                <div
                  class="css-9g59n2-Box e2i8k6n3"
                  color="text"
                  font-size="2"
                >
                  Shipping Routes
                </div>
                <div
                  class="css-11jaf6e-Box e2i8k6n3"
                  color="muted"
                  font-size="0"
                >
                  $
                  0.05
                   per record
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="css-p155cz-Box-Flex e2i8k6n2"
            >
              <img
                alt="dataset thumbnail"
                class="css-p7u6kb-Img e2i8k6n4"
                height="1"
                src="https://picsum.photos/id/1026/92/92"
              />
              <div
                class="css-d9aoz6-Box-Flex e2i8k6n2"
              >
                <div
                  class="css-9g59n2-Box e2i8k6n3"
                  color="text"
                  font-size="2"
                >
                  Ports of Entry
                </div>
                <div
                  class="css-11jaf6e-Box e2i8k6n3"
                  color="muted"
                  font-size="0"
                >
                  $
                  0.02
                   per record
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="css-p155cz-Box-Flex e2i8k6n2"
            >
              <img
                alt="dataset thumbnail"
                class="css-p7u6kb-Img e2i8k6n4"
                height="1"
                src="https://picsum.photos/id/1024/92/92"
              />
              <div
                class="css-d9aoz6-Box-Flex e2i8k6n2"
              >
                <div
                  class="css-9g59n2-Box e2i8k6n3"
                  color="text"
                  font-size="2"
                >
                  Ocean Measurements
                </div>
                <div
                  class="css-11jaf6e-Box e2i8k6n3"
                  color="muted"
                  font-size="0"
                >
                  $
                  0.01
                   per record
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="css-svkabz-Box-Flex e2i8k6n2"
            >
              <img
                alt="dataset thumbnail"
                class="css-p7u6kb-Img e2i8k6n4"
                height="1"
                src="https://picsum.photos/id/135/92/92"
              />
              <div
                class="css-d9aoz6-Box-Flex e2i8k6n2"
              >
                <div
                  class="css-9g59n2-Box e2i8k6n3"
                  color="text"
                  font-size="2"
                >
                  Weather and Climate
                </div>
                <div
                  class="css-11jaf6e-Box e2i8k6n3"
                  color="muted"
                  font-size="0"
                >
                  $
                  0.04
                   per record
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1x3sxzl-Box e2i8k6n3"
      >
        <h3
          class="css-5jl3ge-Box e2i8k6n3"
          color="muted"
          font-size="1"
          font-weight="normal"
          text-decoration="underline"
        >
          Included countries
        </h3>
        <div
          class="css-vq3ac2-Box-Flex e2i8k6n2"
        >
          <div>
            <button
              class="css-1fbado3-Box e2i8k6n3"
              color="text"
              font-size="0"
              height="1"
            >
              United States
            </button>
          </div>
          <div>
            <button
              class="css-1fbado3-Box e2i8k6n3"
              color="text"
              font-size="0"
              height="1"
            >
              United Kingdom
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-sbc3pv-Box e2i8k6n3"
      />
      <div
        class="css-9wihcd-Box-Flex e2i8k6n2"
      >
        <div
          class="css-1x3sxzl-Box e2i8k6n3"
        >
          <button
            class="css-19fpo8m-Box e2i8k6n3"
            cursor="pointer"
            height="100%"
          >
            <div
              class="css-byy0dv-Box e2i8k6n3"
              color="muted"
            >
              Save
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
